### PR TITLE
Fix nightly CI errors

### PR DESCRIPTION
**Pull request by SergioGasquez**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
Added a bunch of allow attributes to get nightly CI green. We might want to address some of those in the future.

#### Testing
Nightly CI run: https://github.com/SergioGasquez/esp-hal/actions/runs/10633769058

### CODE REVIEW

File: esp-wifi/src/common_adapter/mod.rs [hunk 1]
```diff
@@ -311,6 +311,7 @@ mod log {
 // #define ESP_EVENT_DEFINE_BASE(id) esp_event_base_t id = #id
 static mut EVT: i8 = 0;
 #[no_mangle]
+#[allow(unused_unsafe)]
```
Comment (MabezDev) [inline esp-wifi/src/common_adapter/mod.rs:314]: I think we should listen to the lints here, it's not unsafe to create a pointer, only to dereference one.

File: esp-backtrace/src/riscv.rs [hunk 2]
```diff
@@ -12,7 +12,7 @@ pub(super) const RA_OFFSET: usize = 4;
 
 /// Registers saved in trap handler
 #[doc(hidden)]
-#[allow(missing_docs)]
+#[allow(missing_docs, dead_code)]
```
Comment (MabezDev) [inline esp-backtrace/src/riscv.rs:15]: What is the dead code here?

**Review by MabezDev — CHANGES_REQUESTED** [review]

**Comment by bugadani** [comment]
Is this rustc's dead code lint being overactive (again)?

File: esp-wifi/src/common_adapter/mod.rs (hunk 1 above)
Comment (SergioGasquez) [reply esp-wifi/src/common_adapter/mod.rs]: Without the `unsafe` it fails with: 
```
error[E0133]: use of mutable static is unsafe and requires unsafe function or block
   --> src/common_adapter/mod.rs:314:52
    |
314 | static mut WIFI_EVENT: esp_event_base_t = addr_of!(EVT);
    |                                                    ^^^ use of mutable static
    |
    = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior
```

See https://github.com/esp-rs/esp-hal/actions/runs/10366874177/job/28697094152

File: esp-backtrace/src/riscv.rs (hunk 2 above)
Comment (SergioGasquez) [reply esp-backtrace/src/riscv.rs]: Ups, slipped in

File: examples/src/bin/debug_assist.rs [hunk 3]
```diff
@@ -42,7 +42,9 @@ fn main() -> ! {
                 static mut _stack_end: u32;
             }
 
+            #[allow(unused_unsafe)]
```
[the remark this answers is not in the document]
Comment (SergioGasquez) [reply examples/src/bin/debug_assist.rs:45]: I think we will have the same output as in https://github.com/esp-rs/esp-hal/pull/1934#discussion_r1715004692

File: esp-wifi/src/common_adapter/mod.rs (hunk 1 above)
Comment (MabezDev) [reply esp-wifi/src/common_adapter/mod.rs:314]: Interesting, if it is really used why do we need `#[allow(unused_unsafe)]`

File: esp-lp-hal/src/uart.rs [hunk 4]
```diff
@@ -320,6 +320,7 @@ impl embedded_io::Write for LpUart {
             match self.flush_tx() {
                 Ok(_) => break,
                 Err(nb::Error::WouldBlock) => { /* Wait */ }
+                #[allow(unreachable_patterns)]
```
Comment (bugadani) [inline esp-lp-hal/src/uart.rs:323]: Can't we just remove this branch instead?

**Review by bugadani — COMMENTED** [review]
Fun fact: since opening this PR, at least one more lint appeared which fires on a lot of our documentation.

File: esp-wifi/src/common_adapter/mod.rs (hunk 1 above)
Comment (bugadani) [reply esp-wifi/src/common_adapter/mod.rs:314]: I think this is relevant for our use case: https://github.com/rust-lang/rust/issues/129578#issuecomment-2309372169 and the recommendation is to add the `allow` if I read it correctly.

BTW this also should (IMO) be a mut pointer which it somehow isn't in esp-wifi-sys.

File: esp-backtrace/src/riscv.rs [hunk 5]
```diff
@@ -15,6 +15,7 @@ pub(super) const RA_OFFSET: usize = 4;
 #[derive(Default, Clone, Copy)]
 #[cfg_attr(feature = "defmt", derive(defmt::Format))]
 #[repr(C)]
+#[allow(dead_code)]
```
Comment (bugadani) [inline esp-backtrace/src/riscv.rs:18]: I'd do https://github.com/esp-rs/esp-hal/pull/2034/files#diff-7a0430174918ac5644e44736369b62b0298ea3c14c94e8fe409afd4af77786c6R18 instead, what do you think?

Comment (SergioGasquez) [reply esp-backtrace/src/riscv.rs]: Much better :+1:

Comment (bugadani) [reply esp-backtrace/src/riscv.rs]: You'll have to gate the Debug implementation too :)

File: esp-wifi/src/common_adapter/mod.rs (hunk 1 above)
Comment (MabezDev) [reply esp-wifi/src/common_adapter/mod.rs:314]: Ah nice find, bit annoying but if that's the recommendation then so be it.

**Review by bugadani — APPROVED** [review]

**Review by bjoernQ — APPROVED** [review]
LGTM

**Pull request merged**